integrations/python/pymongo/workload-executor.py

```diff
@@ -35,92 +22,66 @@ def interrupt_handler(signum, frame):
-        # TODO: remove this once BUILD-10841 is done.
-        import certifi
-        return MongoClient(mongodb_uri, tlsCAFile=certifi.where())
-    return MongoClient(mongodb_uri)
+def filter_failures_errors(entity_map):
+    e, f = [], []
+    pre_filter_arr = []
+    for entity_type in ["failures", "errors"]:
+        if entity_type in entity_map:
+            pre_filter_arr.extend(entity_map[entity_type])
+    for entity in pre_filter_arr:
+        (e, f)[entity["type"] == AssertionError].append(entity)
+    entity_map._entities["failures"], entity_map._entities["errors"] = f, e
 
 
 def workload_runner(mongodb_uri, test_workload):
-    # Do not modify connection string and do not add any extra options.
-    client = connect(mongodb_uri)
-
-    # Create test entities.
-    database = client.get_database(test_workload["database"])
-    collection = database.get_collection(test_workload["collection"])
-    objects = {"database": database, "collection": collection}
-
-    # Run operations
-    num_failures = 0
-    num_errors = 0
-    num_operations = 0
-    global IS_INTERRUPTED
-
-    operations = test_workload["operations"]
-    ops = [prepare_operation(op) for op in operations]
-
-    while True:
-        if IS_INTERRUPTED:
-            break
-        for op in ops:
-            try:
-                run_operation(objects, op)
-            except AssertionError:
-                traceback.print_exc()
-                num_failures += 1
-            except Exception:
-                traceback.print_exc()
-                num_errors += 1
-            else:
-                num_operations += 1
-
-    # We reach here once IS_INTERRUPTED has been set to True.
-    stats = {"numErrors": num_errors, "numFailures": num_failures,
-             "numSuccesses": num_operations}
-    print("Workload statistics: {!r}".format(stats))
-
-    sentinel = os.path.join(os.path.abspath(os.curdir), 'results.json')
-    print("Writing statistics to sentinel file {!r}".format(sentinel))
-    with open('results.json', 'w') as fr:
-        json.dump(stats, fr)
-    exit(0 or num_errors or num_failures)
+    runner = UnifiedSpecTestMixinV1()
+    runner.TEST_SPEC = test_workload
+    UnifiedSpecTestMixinV1.TEST_SPEC = test_workload
+    runner.setUpClass()
+    runner.setUp()
+    # should be removed in final version, but useful for testing right now
+    for op in test_workload["tests"][0]["operations"]:
+        if op.get("name") == "loop":
+            op["arguments"]["numIterations"] = 10
+    try:
+        runner.run_scenario(test_workload["tests"][0], uri=mongodb_uri)
+    except Exception as exc:
+        target = "failures" if isinstance(exc, AssertionError) else "errors"
+        if target not in runner.entity_map:
+            runner.entity_map[target] = []
+        runner.entity_map[target].append({
+            "error": str(exc),
+            "time": time.time(),
+            "type": type(exc)
+        })
```

Comment:
Actually I don't think that that test proves my point. However, there is a section of the pseudocode implementation that I believe is what inspired my approach 
``` 
try {
        runner.executeScenario();
    } catch (propagatedError) {
        /* If the test runner propagates an error or failure (e.g. it is not
         * captured by the loop or occurs outside of the loop), it MUST be
         * reported by the workload executor. */
         errors.push({
           error: propagatedError.message,
           time: Date.now() / 1000
         });
    }
```

Reply:
It also states this about error propagation:
"Note that is possible for some or all of these arrays to be empty if the corresponding data was not reported by the unified test runner and the test runner did not propagate an error or failure (which would then be reported by the workload executor)."

Reply:
Also I tried it out without the error catching code and the spec says we MUST write out results.json and events.json, but there are assertion errors not being caught by pymongo that cause the program to exit without writing statistics.

Reply:
An example of a test that causes the workload executor to throw an error (rather than it being caught in the loop) is /tests/validator-numFailures-not-captured.yml